### PR TITLE
Emit NAT updates

### PR DESCRIPTION
**Pull request by LuKks**
Closes https://github.com/mafintosh/dht-rpc/issues/65

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -654,7 +658,7 @@ class DHT extends EventEmitter {
     for (const res of pongs) {
       if (hosts.indexOf(res.from.host) > -1) {
         count++
-        natSampler.add(res.to.host, res.to.port)
+        natSampler.add(res.to.host, res.to.port) // + it could be useful to report this one
```
Comment (LuKks) [inline index.js:661]: This is on `_checkIfFirewalled`, but this method being executed doesn't mean that we're going to use that new `natSampler`

**Comment by LuKks** [comment]
@mafintosh Ready!

File: index.js [hunk 2]
```diff
@@ -267,13 +267,28 @@ class DHT extends EventEmitter {
     return req
   }
 
+  _natAdd (host, port) {
+    const prevHost = this.host
+    const prevPort = this.port
```
Comment (mafintosh) [inline index.js:272]: use this._nat.host/port instead

File: index.js [hunk 3]
```diff
@@ -267,13 +267,28 @@ class DHT extends EventEmitter {
...
+
+    this._nat.add(host, port)
+
+    this._natChange(prevHost, prevPort)
```
Comment (mafintosh) [inline index.js:276]: i'd inline this and make it a negative check instead, ie

```
if  (prevHost === this._nat.host && prevPort === this._nat.port) return
```

File: index.js [hunk 4]
```diff
@@ -574,8 +589,13 @@ class DHT extends EventEmitter {
     }
 
     if (natSampler !== this._nat) {
+      const prevHost = this.host
+      const prevPort = this.port
+
       this._nonePersistentSamples = []
       this._nat = natSampler
+
+      this._natChange(prevHost, prevPort)
```
Comment (mafintosh) [inline index.js:598]: would inline it

File: index.js [hunk 5]
```diff
@@ -574,8 +585,15 @@ class DHT extends EventEmitter {
     }
 
     if (natSampler !== this._nat) {
+      const prevHost = this._nat.host
+      const prevPort = this._nat.port
+
       this._nonePersistentSamples = []
       this._nat = natSampler
+
+      if (!(prevHost === this._nat.host && prevPort === this._nat.port)) {
```
Comment (mafintosh) [inline index.js:594]: `!== || !==` is fine here

**Review by kasperisager — APPROVED** [review]
Looks good to me!

**Pull request merged**